dcore/state.py

```diff
@@ -71,24 +77,44 @@ def dump(self):
             self.output.dumplist = self.fieldlist
 
         funcs = self.xn.split()
+        field_dict = {name: func for (name, func) in zip(self.fieldlist, funcs)}
         to_dump = []
-        for name, f in zip(self.fieldlist, funcs):
+        for name, f in field_dict.iteritems():
             if name in self.output.dumplist:
                 to_dump.append(f)
             f.rename(name=name)
 
-        dumpdir = path.join("results", self.output.dirname)
+        if self.output.steady_state_dump_err:
+            init_funcs = self.x_init.split()
+            for name, f, f_init in zip(self.fieldlist, funcs, init_funcs):
+                if name in self.output.dumplist:
+                    err = Function(f.function_space(), name=name+'err').assign(f-f_init)
+                    field_dict[name+"err"] = err
+                    self.diagnostics.register(name+"err")
```

Comment:
I need it for viz too at the moment but I'll bear that in mind, thanks.


________________________________
From: Lawrence Mitchell <notifications@github.com>
Sent: 28 April 2016 12:07
To: firedrakeproject/dcore
Cc: Shipton, Jemma; Author
Subject: Re: [firedrakeproject/dcore] Linear sw (#32)


In dcore/state.py<https://github.com/firedrakeproject/dcore/pull/32#discussion_r61409560>:

>              if name in self.output.dumplist:
>                  to_dump.append(f)
>              f.rename(name=name)
>
> -        dumpdir = path.join("results", self.output.dirname)
> +        if self.output.steady_state_dump_err:
> +            init_funcs = self.x_init.split()
> +            for name, f, f_init in zip(self.fieldlist, funcs, init_funcs):
> +                if name in self.output.dumplist:
> +                    err = Function(f.function_space(), name=name+'err').assign(f-f_init)
> +                    field_dict[name+"err"] = err
> +                    self.diagnostics.register(name+"err")


Note that you don't actually need to allocate a new function for this guy if you're only going to use it in the diagnostics. It suffices to put f - f_init in field_dict. Because when you compute the l2 error you'll end up inserting assemble(sqrt(dot(f - f_init, f-f_init))*dx) which works fine. If you're also going to output it for viz, you can't do this trick though.

-
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub<https://github.com/firedrakeproject/dcore/pull/32/files/1d40d5228193e690628001d7768d188ac091bafb#r61409560>